### PR TITLE
ASoC: SOF: Intel: byt: Add PM callbacks

### PR DESCRIPTION
**Pull request by eballetbo**
This PR is a partial rework of the PR done by @ranj063 some time ago [1]. The purpose of this patch is to simply add PM support for BYT/CHT. Note that as discussed in [1] PM runtime support has no sense for these devices so only suspend/resume support is implemented.

Note also that this patch depends on [2] to work properly. Both PR have been tested on a CHT/BYT device.

[1] https://github.com/thesofproject/linux/pull/1523
[2]  https://github.com/thesofproject/linux/pull/2092

### CODE REVIEW

File: sound/soc/sof/intel/byt.c [hunk 1]
```diff
@@ -455,6 +455,36 @@ static void byt_set_mach_params(const struct snd_soc_acpi_mach *mach,
 	mach_params->platform = dev_name(dev);
 }
 
+static int byt_reset_dsp_disable_int(struct snd_sof_dev *sdev)
```
Comment (cujomalainey) [inline sound/soc/sof/intel/byt.c]: why return an int if the function can never fail?

Comment (lyakh) [reply sound/soc/sof/intel/byt.c]: +1

Comment (eballetbo) [reply sound/soc/sof/intel/byt.c]: pushed a new branch with this change

**Comment by ranj063** [comment]
@eballetbo I shouldnt be approving my own code haha. Lets wait for the others to review.

**Comment by plbossart** [comment]
@eballetbo do you mind clarifying how you tested this and on what platform?
This fails for me on a Cyan Chromebook with the legacy BIOS when suspending while playing - with rtcwake -m mem -s 4. I can however start playing again, it's not blocking somehow.

````
[   96.906960] PM: suspend entry (deep)
[   96.907329] Filesystems sync: 0.000 seconds
[   96.908857] Freezing user space processes ... (elapsed 0.001 seconds) done.
[   96.910340] OOM killer disabled.
[   96.910345] Freezing remaining freezable tasks ... (elapsed 0.001 seconds) done.
[   96.911595] printk: Suspending console(s) (use no_console_suspend to debug)
[   97.006922] sof-audio-acpi 808622A8:00: error: no reply expected, received 0x0
[   97.397851] ACPI: EC: interrupt blocked
[   97.398105] ACPI: Preparing to enter system sleep state S3
[   97.398275] ACPI: EC: event blocked
[   97.398278] ACPI: EC: EC stopped
[   97.398283] PM: Saving platform NVS memory
[   97.398289] Disabling non-boot CPUs ...
[   97.401392] smpboot: CPU 1 is now offline
[   97.404951] smpboot: CPU 2 is now offline
[   97.408578] smpboot: CPU 3 is now offline
[   97.410337] ACPI: Low-level resume complete
[   97.410411] ACPI: EC: EC started
[   97.410412] PM: Restoring platform NVS memory
[   97.414133] Enabling non-boot CPUs ...
[   97.414191] x86: Booting SMP configuration:
[   97.414193] smpboot: Booting Node 0 Processor 1 APIC 0x2
[   97.414839] ACPI: \_PR_.CP01: Found 3 idle states
[   97.415101] CPU1 is up
[   97.415138] smpboot: Booting Node 0 Processor 2 APIC 0x4
[   97.418818] ACPI: \_PR_.CP02: Found 3 idle states
[   97.419074] CPU2 is up
[   97.419103] smpboot: Booting Node 0 Processor 3 APIC 0x6
[   97.419752] ACPI: \_PR_.CP03: Found 3 idle states
[   97.420398] CPU3 is up
[   97.420770] ACPI: Waking up from system sleep state S3
[   97.420969] ACPI: EC: interrupt unblocked
[   97.503510] ACPI: EC: event unblocked
[   99.092916] usb 1-4: reset high-speed USB device number 2 using xhci_hcd
[   99.340677] OOM killer enabled.
[   99.340681] Restarting tasks ... done.
[   99.375904] PM: suspend exit
[  100.354001] sof-audio-acpi 808622A8:00: error: ipc error for 0x60050000 size 12
[  100.354030]  Low Latency: ASoC: trigger FE cmd: 0 failed: -22
````

**Comment by plbossart** [comment]
@eballetbo @ranj063 same problem on an UEFI CHT device
[dmesg_cht.txt](https://github.com/thesofproject/linux/files/4618531/dmesg_cht.txt)

**Comment by eballetbo** [comment]
@plbossart when you say Legacy BIOS you mean SeaBIOS?

I am testing on a Cyan Chromebook with vendor BIOS (not SeaBIOS). Basically the test I did is:
1. Check that audio works `paplay sample.wav`
2. Trigger a suspend with `systemctl suspend`
3. Wake-up by pressing a key
4. Play again an audio file `paplay sample.wav`

Without the patch playing after resume doesn't work, after the patch it work.

I'll test your test-case when I have the device available this afternoon.

**Comment by plbossart** [comment]
> @plbossart when you say Legacy BIOS you mean SeaBIOS?
> 
> I am testing on a Cyan Chromebook with vendor BIOS (not SeaBIOS). Basically the test I did is:
> 
> 1. Check that audio works `paplay sample.wav`
> 2. Trigger a suspend with `systemctl suspend`
> 3. Wake-up by pressing a key
> 4. Play again an audio file `paplay sample.wav`
> 
> Without the patch playing after resume doesn't work, after the patch it work.
> 
> I'll test your test-case when I have the device available this afternoon.

Yes, Control-L -> SeaBIOS.

This happens when you suspend *during* playback.  
on one terminal I have speaker-test -Dhw:2,0 -c2 -r48000
and on the other rtcwake -m mem -s 4

**Comment by eballetbo** [comment]
> > @plbossart when you say Legacy BIOS you mean SeaBIOS?
> > I am testing on a Cyan Chromebook with vendor BIOS (not SeaBIOS). Basically the test I did is:
> > 
> > 1. Check that audio works `paplay sample.wav`
> > 2. Trigger a suspend with `systemctl suspend`
> > 3. Wake-up by pressing a key
> > 4. Play again an audio file `paplay sample.wav`
> > 
> > Without the patch playing after resume doesn't work, after the patch it work.
> > I'll test your test-case when I have the device available this afternoon.
> 
> Yes, Control-L -> SeaBIOS.
> 
> This happens when you suspend _during_ playback.
> on one terminal I have speaker-test -Dhw:2,0 -c2 -r48000
> and on the other rtcwake -m mem -s 4

Yes, I can confirm the issue, suspend/resume while playing doesn't work.

**Comment by ranj063** [comment]
> @eballetbo @ranj063 same problem on an UEFI CHT device
> [dmesg_cht.txt](https://github.com/thesofproject/linux/files/4618531/dmesg_cht.txt)

@plbossart I think I know why this is happening. Let me send you a patch for this in a bit

**Comment by ranj063** [comment]
@plbossart @eballetbo Could you please try this change to see if it fixes your issue? Basically since we do not meddle with the DSP states for BYT/CHT, the old state remains in D0 and we end up skipping the FW reloading process upon resuming.
I am still a bit sceptical if this would fix the issue entirely because I am not sure whether the DSP is actually powered off or not during suspend. But worth a try I think

```
diff --git a/sound/soc/sof/pm.c b/sound/soc/sof/pm.c
index ce687fe69924..a5f7c7f024a1 100644
--- a/sound/soc/sof/pm.c
+++ b/sound/soc/sof/pm.c
@@ -114,8 +114,12 @@ static int sof_resume(struct device *dev, bool runtime_resume)
                return ret;
        }
 
-       /* Nothing further to do if resuming from a low-power D0 substate */
-       if (!runtime_resume && old_state == SOF_DSP_PM_D0)
+       /*
+        * Nothing further to be done for platforms that support the low power
+        * D0 substate.
+        */
+       if (!runtime_resume && sof_ops(sdev)->set_power_state &&
+           old_state == SOF_DSP_PM_D0)
                return 0;
 
        sdev->fw_state = SOF_FW_BOOT_PREPARE;
```

**Comment by eballetbo** [comment]
> @plbossart @eballetbo Could you please try this change to see if it fixes your issue? Basically since we do not meddle with the DSP states for BYT/CHT, the old state remains in D0 and we end up skipping the FW reloading process upon resuming.
> I am still a bit sceptical if this would fix the issue entirely because I am not sure whether the DSP is actually powered off or not during suspend. But worth a try I think
> 
> ```
> diff --git a/sound/soc/sof/pm.c b/sound/soc/sof/pm.c
> index ce687fe69924..a5f7c7f024a1 100644
> --- a/sound/soc/sof/pm.c
> +++ b/sound/soc/sof/pm.c
> @@ -114,8 +114,12 @@ static int sof_resume(struct device *dev, bool runtime_resume)
>                 return ret;
>         }
>  
> -       /* Nothing further to do if resuming from a low-power D0 substate */
> -       if (!runtime_resume && old_state == SOF_DSP_PM_D0)
> +       /*
> +        * Nothing further to be done for platforms that support the low power
> +        * D0 substate.
> +        */
> +       if (!runtime_resume && sof_ops(sdev)->set_power_state &&
> +           old_state == SOF_DSP_PM_D0)
>                 return 0;
>  
>         sdev->fw_state = SOF_FW_BOOT_PREPARE;
> ```

Doesn't seem to solve the problem, I got the same issue

Hmm, note that on the patch I added the line below in byt_suspend It sets dsp_power_state to D3? But that's maybe a workaround, but didn't work without it.

```
sdev->dsp_power_state.state = target_state;
```

Maybe I messed up the power state ...

**Comment by ranj063** [comment]
> Hmm, note that on the patch I added the line below in byt_suspend It sets dsp_power_state to D3? But that's maybe a workaround, but didn't work without it.
> 
> ```
> sdev->dsp_power_state.state = target_state;
> ```
> 
> Maybe I messed up the power state ...

@eballetbo this was a local change? I dont see this change in the PR? 
But we cant set the power state for BYT because the driver doesnt really power down the DSP.

File: sound/soc/sof/intel/byt.c [hunk 2]
```diff
@@ -455,6 +455,36 @@ static void byt_set_mach_params(const struct snd_soc_acpi_mach *mach,
...
+
+	/* Put DSP into reset, set reset vector */
+	snd_sof_dsp_update_bits64(sdev, BYT_DSP_BAR, SHIM_CSR,
+				  SHIM_BYT_CSR_RST | SHIM_BYT_CSR_VECTOR_SEL,
+				  SHIM_BYT_CSR_RST | SHIM_BYT_CSR_VECTOR_SEL);
+}
+
+static int byt_suspend(struct snd_sof_dev *sdev, u32 target_state)
+{
+	sdev->dsp_power_state.state = target_state;
```
Comment (eballetbo) [inline sound/soc/sof/intel/byt.c]: @ranj063 Here ^(the original patch didn't have this)

**Comment by eballetbo** [comment]
> > Hmm, note that on the patch I added the line below in byt_suspend It sets dsp_power_state to D3? But that's maybe a workaround, but didn't work without it.
> > ```
> > sdev->dsp_power_state.state = target_state;
> > ```
> > 
> > 
> > Maybe I messed up the power state ...
> 
> @eballetbo this was a local change? I dont see this change in the PR?

It is: https://github.com/thesofproject/linux/pull/2094/commits/c40a7c449946e3334ef69090be6ced2004a99621#r424597157

> But we cant set the power state for BYT because the driver doesnt really power down the DSP.

**Comment by plbossart** [comment]
> Maybe I messed up the power state ...

I think there's indeed something wrong with the state machine, if you look at my dmesg log I can see 

````
[   37.550669] sof-audio-acpi 808622A8:00: ipc tx: 0x40020000: GLB_PM_MSG: CTX_RESTORE
[   37.550725] sof-audio-acpi 808622A8:00: ipc tx succeeded: 0x40020000: GLB_PM_MSG: CTX_RESTORE
[   37.551652] sof-audio-acpi 808622A8:00: ipc rx: 0x90020000: GLB_TRACE_MSG
[   37.551671] sof-audio-acpi 808622A8:00: ipc rx done: 0x90020000: GLB_TRACE_MSG
[   37.588199] r8169 0000:02:00.0 enp2s0: Link is Down
[   37.779035] OOM killer enabled.
[   37.779040] Restarting tasks ... 
[   37.782687] sof-audio-acpi 808622A8:00: PCM: stream 0 dir 0 DMA position 8260 DAI position 737320
[   37.783238] sof-audio-acpi 808622A8:00: PCM: stream 0 dir 0 DMA position 8260 DAI position 737320
[   37.784248] done.
[   37.784613] video LNXVIDEO:00: Restoring backlight state
[   37.784828] PM: suspend exit
[   38.288181] sof-audio-acpi 808622A8:00: PCM: stream 0 dir 0 DMA position 8260 DAI position 737320
[   38.792253] sof-audio-acpi 808622A8:00: pcm: trigger stream 0 dir 0 cmd 0
[   38.792263] sof-audio-acpi 808622A8:00: ipc tx: 0x60050000: GLB_STREAM_MSG: TRIG_STOP
[   38.792357] sof-audio-acpi 808622A8:00: error: ipc error for 0x60050000 size 12
[   38.792364]  Low Latency: ASoC: trigger FE cmd: 0 failed: -22
[   38.792377] sof-audio-acpi 808622A8:00: pcm: free stream 0 dir 0
[   38.792380] sof-audio-acpi 808622A8:00: ipc tx: 0x60030000: GLB_STREAM_MSG: PCM_FREE
[   38.792459] sof-audio-acpi 808622A8:00: ipc tx succeeded: 0x60030000: GLB_STREAM_MSG: PCM_FREE
````
So it looks like we restart but end-up with an underflow, which will result in a stop but that stop isn't valid somehow. In other words maybe there is a missing start.

File: sound/soc/sof/intel/byt.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/intel/byt.c]: @eballetbo sorry I dont know how I missed this. we should remove it but it shouldnt make a difference anyway.

**Comment by ranj063** [comment]
> So it looks like we restart but end-up with an underflow, which will result in a stop but that stop isn't valid somehow. In other words maybe there is a missing start.

@plbossart the other thing missing is that ideally the stream should be stopped when you suspend. I dont see that stop trigger before suspend entry in the dmesg. Do you think it has something to do with the SNDRV_PCM_INFO_RESUME flag?

**Comment by plbossart** [comment]
> > So it looks like we restart but end-up with an underflow, which will result in a stop but that stop isn't valid somehow. In other words maybe there is a missing start.
> 
> @plbossart the other thing missing is that ideally the stream should be stopped when you suspend. I dont see that stop trigger before suspend entry in the dmesg. Do you think it has something to do with the SNDRV_PCM_INFO_RESUME flag?

Possibly, but I don't see RESUME set anywhere for SOF.

**Comment by ranj063** [comment]
> Possibly, but I don't see RESUME set anywhere for SOF.

@plbossart would this fix the issue although I'm not sure if it is accurate to say that SOF supports full resume for these platforms:
```
diff --git a/sound/soc/sof/intel/byt.c b/sound/soc/sof/intel/byt.c
index 37ec979612d1..3a71ca2bce41 100644
--- a/sound/soc/sof/intel/byt.c
+++ b/sound/soc/sof/intel/byt.c
@@ -668,7 +668,8 @@ const struct snd_sof_dsp_ops sof_tng_ops = {
                        SNDRV_PCM_INFO_MMAP_VALID |
                        SNDRV_PCM_INFO_INTERLEAVED |
                        SNDRV_PCM_INFO_PAUSE |
-                       SNDRV_PCM_INFO_BATCH,
+                       SNDRV_PCM_INFO_BATCH |
+                       SNDRV_PCM_INFO_RESUME,
 
        .arch_ops = &sof_xtensa_arch_ops,
 };
@@ -865,7 +866,8 @@ const struct snd_sof_dsp_ops sof_byt_ops = {
                        SNDRV_PCM_INFO_MMAP_VALID |
                        SNDRV_PCM_INFO_INTERLEAVED |
                        SNDRV_PCM_INFO_PAUSE |
-                       SNDRV_PCM_INFO_BATCH,
+                       SNDRV_PCM_INFO_BATCH |
+                       SNDRV_PCM_INFO_RESUME,
 
        .arch_ops = &sof_xtensa_arch_ops,
 };
@@ -943,7 +945,8 @@ const struct snd_sof_dsp_ops sof_cht_ops = {
                        SNDRV_PCM_INFO_MMAP_VALID |
                        SNDRV_PCM_INFO_INTERLEAVED |
                        SNDRV_PCM_INFO_PAUSE |
-                       SNDRV_PCM_INFO_BATCH,
+                       SNDRV_PCM_INFO_BATCH |
+                       SNDRV_PCM_INFO_RESUME,
 
        .arch_ops = &sof_xtensa_arch_ops,
 };
```

**Comment by plbossart** [comment]
> > Possibly, but I don't see RESUME set anywhere for SOF.
> 
> @plbossart would this fix the issue although I'm not sure if it is accurate to say that SOF supports full resume for these platforms:
> 
> ```
> diff --git a/sound/soc/sof/intel/byt.c b/sound/soc/sof/intel/byt.c
> index 37ec979612d1..3a71ca2bce41 100644
> --- a/sound/soc/sof/intel/byt.c
> +++ b/sound/soc/sof/intel/byt.c
> @@ -668,7 +668,8 @@ const struct snd_sof_dsp_ops sof_tng_ops = {
>                         SNDRV_PCM_INFO_MMAP_VALID |
>                         SNDRV_PCM_INFO_INTERLEAVED |
>                         SNDRV_PCM_INFO_PAUSE |
> -                       SNDRV_PCM_INFO_BATCH,
> +                       SNDRV_PCM_INFO_BATCH |
> +                       SNDRV_PCM_INFO_RESUME,
>  
>         .arch_ops = &sof_xtensa_arch_ops,
>  };
> @@ -865,7 +866,8 @@ const struct snd_sof_dsp_ops sof_byt_ops = {
>                         SNDRV_PCM_INFO_MMAP_VALID |
>                         SNDRV_PCM_INFO_INTERLEAVED |
>                         SNDRV_PCM_INFO_PAUSE |
> -                       SNDRV_PCM_INFO_BATCH,
> +                       SNDRV_PCM_INFO_BATCH |
> +                       SNDRV_PCM_INFO_RESUME,
>  
>         .arch_ops = &sof_xtensa_arch_ops,
>  };
> @@ -943,7 +945,8 @@ const struct snd_sof_dsp_ops sof_cht_ops = {
>                         SNDRV_PCM_INFO_MMAP_VALID |
>                         SNDRV_PCM_INFO_INTERLEAVED |
>                         SNDRV_PCM_INFO_PAUSE |
> -                       SNDRV_PCM_INFO_BATCH,
> +                       SNDRV_PCM_INFO_BATCH |
> +                       SNDRV_PCM_INFO_RESUME,
>  
>         .arch_ops = &sof_xtensa_arch_ops,
>  };
> ```

looks like a really bad late afternoon hack to push the issue under the rug... I think there's something else here.

**Comment by eballetbo** [comment]
Just to confirm that the above patch doesn't fix the issue. Definitely there is something else here.

File: sound/soc/sof/intel/byt.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/intel/byt.c]: @ranj063 it does make a difference, big time! I cannot restart playing after a system suspend/resume transition if I comment out this line. I think these patch series predates all the S0ix rework and there are some gaps in the state machines.

@eballetbo if you can confirm my finding it'd be nice, thanks!

Comment (eballetbo) [reply sound/soc/sof/intel/byt.c]: You should remove that line but apply what @ranj063 proposed instead. 
* https://github.com/thesofproject/linux/pull/2094#issuecomment-628110819
Then works as before, suspend/resume works but suspend/resume while playing no. I can upload and updated patch

Comment (plbossart) [reply sound/soc/sof/intel/byt.c]: Thanks, yes please add this patch so that we don' need to track review comments ;-)

**Comment by eballetbo** [comment]
What I see is what @ranj063 pointed, TRIG_STOP is not issued on suspend, neither TRIG_START on resume. But is not still clear to me whether should be issued or not, and the the answer is yes, not sure if should be specific to byt driver or in the core. I suspect is the core?

Sorry for dummy question but I'm just introducing to this subsystem.

File: sound/soc/sof/intel/byt.c [hunk 3]
```diff
@@ -428,6 +428,34 @@ static void byt_set_mach_params(const struct snd_soc_acpi_mach *mach,
 	mach_params->platform = dev_name(dev);
```
Comment (dbaluta) [inline sound/soc/sof/intel/byt.c:428]: @ranj063 @eballetbo this patch does more then adding PM callbacks. Care to add more information in the commit message (or even if possible split this in two separate patches).
I am also now looking into PM for i.MX.

**Comment by plbossart** [comment]
@eballetbo @ranj063 I got it, we don't have the .pm_ops in machine drivers.... Gah. will submit another PR based on this one.

**Comment by ranj063** [comment]
> @eballetbo @ranj063 I got it, we don't have the .pm_ops in machine drivers.... Gah. will submit another PR based on this one.

Nice catch. That makes sense now

**Comment by plbossart** [comment]
replaced by PR #2138. @eballetbo if you don't mind checking it'd be great!
